File: HyeKyungKo/2022-10-28-Merge-Two-Sorted-Lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

//Time Complexity: O(n+m) <- n: size of list1, m : size of list2
//Space Complexity: O(1)
class Solution{
    public ListNode mergeTwoLists(ListNode list1, ListNode list2){
        
        ListNode head = new ListNode();
        ListNode merged = head;
        while(list1 != null || list2 != null){
            if(list1 == null){
                merged.next = list2;
                break;
            }else if(list2 == null){
                merged.next = list1;
                break;
            }else{
                if(list1.val <= list2.val){
                    merged.next = list1;
                    list1 = list1.next;
                }else{
                    merged.next = list2;
                    list2 = list2.next;
                }
                merged = merged.next;
            }            
        }
        
        return head.next;
    }
}

File: Chris/2022-10-28-Merge Two Sorted Lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        # Check empty lists
        if list1 == None:
            return list2
        if list2 == None:
            return list1
        
        head1, head2 = list1, list2
        
        if head1.val < head2.val:
            res = cur = head1
            head1 = head1.next
        else:
            res = cur = head2
            head2 = head2.next
            
        while head1 and head2:
            if head1.val < head2.val:
                cur.next = head1
                head1 = head1.next
            else:
                cur.next = head2
                head2 = head2.next
            cur = cur.next
        
        if head1 == None:
            cur.next = head2
        if head2 == None:
            cur.next = head1

        
        return res
